src/main/scala/io/qbeast/spark/delta/IndexFiles.scala

```diff
@@ -0,0 +1,199 @@
+   *   the path to the root of the index
+   * @param indexFile
+   *   the IndexFile instance
+   * @return
+   *   a FileStatus instance
+   */
+  def toFileStatus(indexPath: Path)(indexFile: IndexFile): FileStatus = {
+    var path = new Path(new URI(indexFile.path))
+    if (!path.isAbsolute()) {
+      path = new Path(indexPath, path)
+    }
+    new FileStatus(indexFile.size, false, 0, 0, indexFile.modificationTime, path)
+  }
+
+  def toFileStatusWithMetadata(indexPath: Path, metadata: Map[String, String] = Map.empty)(
+      indexFile: IndexFile): FileStatusWithMetadata = {
+    FileStatusWithMetadata(toFileStatus(indexPath)(indexFile), metadata)
+  }
+
+  private def encodeBlocks(blocks: IISeq[Block]): String = {
+    val writer = new StringWriter()
+    val generator = jsonFactory.createGenerator(writer)
+    generator.writeStartArray()
+    blocks.foreach { block =>
+      generator.writeStartObject()
+      generator.writeStringField("cubeId", block.cubeId.string)
+      generator.writeNumberField("minWeight", block.minWeight.value)
+      generator.writeNumberField("maxWeight", block.maxWeight.value)
+      generator.writeNumberField("elementCount", block.elementCount)
+      generator.writeBooleanField("replicated", block.replicated)
+      generator.writeEndObject()
+    }
+    generator.writeEndArray()
+    generator.close()
+    writer.close()
+    writer.toString()
+  }
+
+  private def decodeBlocks(json: String, dimensionCount: Int, builder: IndexFileBuilder): Unit = {
+    val parser = jsonFactory.createParser(json)
+    parser.nextToken()
+    if (!parser.hasToken(JsonToken.START_ARRAY)) {
+      throw new JsonParseException(parser, "Array start is expected")
+    }
+    parseBlocks(parser, dimensionCount, builder)
+    parser.close()
+  }
+
+  private def parseBlocks(
+      parser: JsonParser,
+      dimensionCount: Int,
+      builder: IndexFileBuilder): Unit = {
+    parser.nextToken()
+    while (parser.hasToken(JsonToken.START_OBJECT)) {
+      parseBlock(parser, dimensionCount, builder.beginBlock())
+      parser.nextToken()
+    }
+    if (!parser.hasToken(JsonToken.END_ARRAY)) {
+      throw new JsonParseException(parser, "Array end is expected")
+    }
+  }
+
+  private def parseBlock(parser: JsonParser, dimensionCount: Int, builder: BlockBuilder): Unit = {
+    parser.nextToken()
+    while (parser.hasToken(JsonToken.FIELD_NAME)) {
+      parser.nextToken()
+      parser.currentName() match {
+        case "cubeId" => builder.setCubeId(CubeId(dimensionCount, parser.getValueAsString()))
+        case "minWeight" => builder.setMinWeight(Weight(parser.getValueAsInt()))
+        case "maxWeight" => builder.setMaxWeight(Weight(parser.getValueAsInt()))
+        case "elementCount" => builder.setElementCount(parser.getValueAsLong())
+        case "replicated" => builder.setReplicated(parser.getValueAsBoolean())
+        case name => throw new JsonParseException(parser, s"Unexpected field '${name}'")
```

Comment:
Could we please create an issue to add unit test cases that validate that this code throws when expected? Codecov is complaining correctly about lines 189 and 194.

Reply:
Issue #296 opened